virttest/qemu_vm.py

```diff
@@ -1900,6 +1900,15 @@ def sort_key(dev):
                               pci_bus=pci_bus)
                 iov += 1
 
+        # Add vsock device, cid 0-2 are reserved by system
+        default_cid = 3
+        for vsock in params.objects('vsock'):
+            vsock_params = params.object_params(vsock)
+            guest_cid = vsock_params.get("guest_cid", default_cid)
+            devices.insert(qdevices.QDevice(
+                "vhost-vsock-pci", {"id": vsock, "guest-cid": guest_cid}))
+            default_cid += default_cid
```

Comment:
Seems libvirt has a more robust approach to allocate cid, here are some snippets, please have a look, thanks.
```c
int
qemuProcessOpenVhostVsock(virDomainVsockDefPtr vsock)
{
    qemuDomainVsockPrivatePtr priv = (qemuDomainVsockPrivatePtr)vsock->privateData;
    const char *vsock_path = "/dev/vhost-vsock";
    int fd;

    if ((fd = open(vsock_path, O_RDWR)) < 0) {
        virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
                       "%s", _("unable to open vhost-vsock device"));
        return -1;
    }

    if (vsock->auto_cid == VIR_TRISTATE_BOOL_YES) {
        if (virVsockAcquireGuestCid(fd, &vsock->guest_cid) < 0)
            goto error;
    } else {
        if (virVsockSetGuestCid(fd, vsock->guest_cid) < 0)
            goto error;
    }

    priv->vhostfd = fd;
    return 0;

 error:
    VIR_FORCE_CLOSE(fd);
    return -1;
}

#if HAVE_DECL_VHOST_VSOCK_SET_GUEST_CID
static int
virVsockSetGuestCidQuiet(int fd,
                         unsigned int guest_cid)
{
    uint64_t val = guest_cid;

    return ioctl(fd, VHOST_VSOCK_SET_GUEST_CID, &val);
}

#else
static int
virVsockSetGuestCidQuiet(int fd ATTRIBUTE_UNUSED,
                         unsigned int guest_cid ATTRIBUTE_UNUSED)
{
    errno = ENOSYS;
    return -1;
}
#endif


/**
 * virVsockSetGuestCid:
 * @fd: file descriptor of a vsock interface
 * @guest_cid: guest CID to be set
 *
 * Wrapper for VHOST_VSOCK_SET_GUEST_CID ioctl.
 * Returns: 0 on success, -1 on error.
 */
int
virVsockSetGuestCid(int fd,
                    unsigned int guest_cid)
{
    if (virVsockSetGuestCidQuiet(fd, guest_cid) < 0) {
        virReportSystemError(errno, "%s",
                             _("failed to set guest cid"));
        return -1;
    }

    return 0;
}

#define VIR_VSOCK_GUEST_CID_MIN 3

/**
 * virVsockAcquireGuestCid:
 * @fd: file descriptor of a vsock interface
 * @guest_cid: where to store the guest CID
 *
 * Iterates over usable CIDs until a free one is found.
 * Returns: 0 on success, with the acquired CID stored in guest_cid
 *         -1 on error.
 */
int
virVsockAcquireGuestCid(int fd,
                        unsigned int *guest_cid)
{
    unsigned int cid = VIR_VSOCK_GUEST_CID_MIN;

    for (; virVsockSetGuestCidQuiet(fd, cid) < 0; cid++) {
        if (errno != EADDRINUSE) {
            virReportSystemError(errno, "%s",
                                 _("failed to acquire guest cid"));
            return -1;
        }
    }
    *guest_cid = cid;

    return 0;
}
```

Reply:
Okay, I will try to translate it into python.

Reply:
@luckyh Before I go further, since the definition of `VHOST_VSOCK_SET_GUEST_CID` is `#define VHOST_VSOCK_SET_GUEST_CID  _IOW(VHOST_VIRTIO, 0x60, __u64)`, there is the linux macro`_IOW` to calculate the request of  `VHOST_VIRTIO`. So probably we need introduce something like https://python-ioctl.readthedocs.io/en/stable/linux.html#ioctl.linux.IOW to achieve it . Do we want that? Or any other suggestion? Thanks.

Reply:
Simply, we can directly record the target values since they are (usually) stable, see `virttest/arch.py`. And yes, the improvement is like something you mentioned, and I guess it will take more time..

Reply:
@luckyh I don't see a way to directly get the value without _IOW support, it is not different from those values defined in virttest/arch.py, which are all directly defined values. And also there is no such macro defined like sockios.h, to allow use get and set something like if properties. It needs to increase the value each time user call it, so read and write action together(like 1++), so _IOW definitely required.

Reply:
@luckyh Honestly I don't think it worth that for this feature, so I will just keep current implementation it require IOW support, although it needs user define the cid if there is multiple vms use vsock, but I guess we will never have such scenario, since we don't have this device by default and will only have few cases test with it.

Reply:
You can use the following C program to get the value.
```c
#include <stdio.h>
#include <linux/vhost.h>

int main(void) {
    printf("0x%x\n", VHOST_VSOCK_SET_GUEST_CID);
    return 0;
}
```
I have run on few platforms, just take a reference:
* x86_64 (kernel 3.10 & 4.20):  0x4008af60
* ppc64le (kernel 4.19): 0x8008af60

Anyway, it's okay for me if you do think it is not worth for now, but please leave a FIXME comment for that.

Thanks.

Reply:
@luckyh That's not what we want to get, what you get is the value of `VHOST_VSOCK_SET_GUEST_CID`,  it's like the address of the function to set the CID, while what we want is the CID set by it, in libvirt it's `ioctl(fd, VHOST_VSOCK_SET_GUEST_CID, &val)`, so I think that's the `*__u64` from `_IOW(VHOST_VIRTIO, 0x60, __u64)`, which should supposed to be a int(or unsigned long long) we passed to it(>=3), and that's how libvirt works.
And also, that's should be increased, a stationary value would help nothing for multiple vsock CID, still need ++ them in the loop.
The correct way to get it in C is just copy the whole libvirt code. I don't think we should do that, if we do want it, it should be in python, so introduce IOW in python.

Reply:
I'm sorry I don't get your point, doesn't the libvirt's logic be:
```
Open vsock (qemuProcessOpenVhostVsock()):

1. check whether the vsock feature is supported (open(vsock_path, O_RDWR)) < 0)

2. try to occupy cid
2a. auto_cid scenario (if (vsock->auto_cid == VIR_TRISTATE_BOOL_YES))
    2a.1. pick a candidate cid, started by 3 (VIR_VSOCK_GUEST_CID_MIN)

    2a.2. test if the interest cid can be used (ioctl(fd, VHOST_VSOCK_SET_GUEST_CID, &val))
    2a.2a. cid is valid (virVsockSetGuestCidQuiet(fd, cid) >= 0)
        2a.2a.1. end the loop and use that one

    2a.2b. cid is invalid but not in-used (if (errno != EADDRINUSE))
        2a.2b.1. give up to find a valid cid and report error

    2a.2c. cid is invalid due to in-used
        2a.2c.1. try next one for lucks (cid++)

    2a.3. use that verified cid (*guest_cid = cid)

2b. user-defined cid scenario
    2b.1. set cid or report error (virVsockSetGuestCid())

3. assign cid to domain
```

So except `VHOST_VSOCK_SET_GUEST_CID` there is nothing requires `_IOW`, you can change the `ioctl` statement to `ioctl(fd, 0x4008af60, &val)` on x86_64 and it would work, the `&val` in this statement is the cid.

The differences between your current approach and libvirt's are:

1. the libvirt one would do cid verification for both auto_cid and user-defined scenario
2. for auto_cid scenario, the libvirt one would skip those candidates which are in-used (and your approach not)

Reply:
@luckyh My fault, I misunderstood what you mean by 'target values' previously, and also I lead myself to the wrong direction. Certainly we can get the value by ioctl when we have the function address. I will update it. Thank you for bearing with me.